src/core_codemods/combine_isinstance_issubclass.py

```diff
@@ -0,0 +1,103 @@
+
+
+class CombineIsinstanceIssubclass(SimpleCodemod, NameResolutionMixin):
+    metadata = Metadata(
+        name="combine-isinstance-issubclass",
+        summary="Simplify Boolean Expressions Using `isinstance` and `issubclass`",
+        review_guidance=ReviewGuidance.MERGE_WITHOUT_REVIEW,
+        references=[],
+    )
+    change_description = "Use tuple of matches instead of boolean expression with `isinstance` or `issubclass`"
+
+    def leave_BooleanOperation(
+        self, original_node: cst.BooleanOperation, updated_node: cst.BooleanOperation
+    ) -> cst.CSTNode:
+        if not self.filter_by_path_includes_or_excludes(
+            self.node_position(original_node)
+        ):
+            return updated_node
+
+        if self.matches_isinstance_issubclass_or_pattern(original_node):
+            self.report_change(original_node)
+            return self.make_new_call_from_boolean_operation(updated_node)
+
+        return updated_node
+
+    def matches_isinstance_issubclass_or_pattern(
+        self, node: cst.BooleanOperation
+    ) -> bool:
+        # Match the pattern: isinstance(x, cls1) or isinstance(x, cls2) or isinstance(x, cls3) or ...
+        # and the same but with issubclass
+        args = [m.Arg(value=m.Name()), m.Arg(value=m.Name() | m.Tuple())]
+        isinstance_call = m.Call(
+            func=m.Name("isinstance"),
+            args=args,
+        )
+        issubclass_call = m.Call(
+            func=m.Name("issubclass"),
+            args=args,
+        )
+        isinstance_or = m.BooleanOperation(
+            left=isinstance_call, operator=m.Or(), right=isinstance_call
+        )
+        issubclass_or = m.BooleanOperation(
+            left=issubclass_call, operator=m.Or(), right=issubclass_call
+        )
+
+        # Check for simple case: isinstance(x, cls) or issubclass(x, cls)
+        if (
+            m.matches(node, isinstance_or | issubclass_or)
+            and node.left.func.value == node.right.func.value  # Same function
+            and node.left.args[0].value.value
+            == node.right.args[0].value.value  # Same first argument (instance)
+        ):
+            return True
+
+        # Check for chained case: isinstance(x, cls1) or isinstance(x, cls2) or isinstance(x, cls3) or ...
+        chained_or = m.BooleanOperation(
+            left=m.BooleanOperation(operator=m.Or()),
+            operator=m.Or(),
+            right=isinstance_call | issubclass_call,
+        )
+        if m.matches(node, chained_or):
+            return all(
+                (
+                    call.func.value == node.right.func.value  # Same function
+                    and call.args[0].value.value
+                    == node.right.args[0].value.value  # Same first argument (instance)
+                )
+                for call in extract_boolean_operands(node, ensure_type=cst.Call)
+            )
+
+        return False
+
+    def make_new_call_from_boolean_operation(
```

Comment:
alright so now that the other codemod PR was merged, I think you can dedupe this and move to a utility location.

Reply:
I wanted to do this as well but `make_new_call_from_boolean_operation` in `combine_startswith_endswith.py` and `combine_isinstance_issubclass.py` are actually a bit different which is why I didn't generalize and add to `utils.py`:

The differences are:
- `startswith` considers only strings so the `evaluated_value` instead of `value` should be considered to avoid duplicate args if the same arg is passed as both a `SimpleString` and `ConcatenatedString` for example. 
- `startswith` the tuple or arg to combine is `call.args[0].value` For `isinstance` the tuple or arg to combine is `call.args[1].value`.
- `startswith` the instance is part of the Call object. For `isinstance` the instance is `call.args[0].value` (one of the args preceding the combined arg)

To generalize this I could rewrite with kwargs `args_to_insert`, `args_to_combine` and `dedupe_attr` and pass the corresponding values but I thought this would add extra complexity when accounting for arg order, for little reusability unless other codeMods will require this type of logic, then its probably worth it. 

What do you think you?

```python3
#combine_startswith_endswith.py
    def make_new_call_from_boolean_operation(
        self, updated_node: cst.BooleanOperation
    ) -> cst.Call:
        elements = []
        seen_evaluated_values = set()
        for call in extract_boolean_operands(updated_node, ensure_type=cst.Call):
            arg_value = call.args[0].value
            arg_elements = (
                arg_value.elements
                if isinstance(arg_value, cst.Tuple)
                else (cst.Element(value=arg_value),)
            )

            for element in arg_elements:
                if (
                    evaluated_value := getattr(element.value, "evaluated_value", None)
                ) in seen_evaluated_values:
                    # If an element has a non-None evaluated value that has already been seen, continue to avoid duplicates
                    continue
                if evaluated_value is not None:
                    seen_evaluated_values.add(evaluated_value)
                elements.append(element)

        new_arg = cst.Arg(value=cst.Tuple(elements=elements))
        return cst.Call(func=call.func, args=[new_arg])
```

```python3
#combine_isinstance_issubclass.py
    def make_new_call_from_boolean_operation(
        self, updated_node: cst.BooleanOperation
    ) -> cst.Call:
        elements = []
        seen_values = set()
        for call in extract_boolean_operands(updated_node, ensure_type=cst.Call):
            class_or_tuple_arg_value = call.args[1].value
            arg_elements = (
                class_or_tuple_arg_value.elements
                if isinstance(class_or_tuple_arg_value, cst.Tuple)
                else (cst.Element(value=class_or_tuple_arg_value),)
            )

            for element in arg_elements:
                if (value := getattr(element.value, "value", None)) in seen_values:
                    # If an element has a non-None evaluated value that has already been seen, continue to avoid duplicates
                    continue
                if value is not None:
                    seen_values.add(value)
                elements.append(element)

        instance_arg = updated_node.left.args[0]
        new_class_or_tuple_arg = cst.Arg(value=cst.Tuple(elements=elements))
        return cst.Call(func=call.func, args=[instance_arg, new_class_or_tuple_arg])
```

Reply:
Yes I think you have the right idea. Here's what I came up with. YOu may need to tweak as I didn't test it

```
def make_new_call_from_boolean_operation(self, updated_node: cst.BooleanOperation, arg_index: int, attribute: str, use_original_arg=False) -> cst.Call:
    elements = []
    seen_values = set()
    for call in extract_boolean_operands(updated_node, ensure_type=cst.Call):
        arg_value = call.args[arg_index].value
        arg_elements = (
            arg_value.elements
            if isinstance(arg_value, cst.Tuple)
            else (cst.Element(value=arg_value),)
        )
        for element in arg_elements:
            unique_value = getattr(element.value, attribute, None)
            if unique_value in seen_values:
                continue
            if unique_value is not None:
                seen_values.add(unique_value)
            elements.append(element)

    new_arg = cst.Arg(value=cst.Tuple(elements=elements))
    new_args = [updated_node.left.args[0], new_arg] if use_original_arg else [new_arg]
    return cst.Call(func=call.func, args=new_args)
```

Reply:
Awesome this will definitely work, thank you. I'll have some time to work on this later today

I also was thinking while testing Pixee last night that both codeMods could be even better by supporting more complex boolean operations. Previous implementations required a series of ors of the same call to be merged, but it could support merging 2 or more of the same call anywhere in a boolean operation chain, regardless of the operands before or after the mergeable calls.

For example:
```python3
# before 
x and x.startswith('foo') or x.startswith('bar')
# after
x and x.startswith(('foo', 'bar'))
```
I have a few ideas for how to implement this while maintaining current behavior and I'll test it with something like this. 
```python3
    @each_func
    @pytest.mark.parametrize(
        "input_code, expected",
        [
            # same name and/or on left            
            ("x and x.{func}('foo') or x.{func}('bar')", "x and x.{func}(('foo', 'bar'))"),
            ("x or x.{func}('foo') or x.{func}('bar')", "x or x.{func}(('foo', 'bar'))"),

            # same name and/or on right
            ("x.{func}('foo') or x.{func}('bar') and x", "x.{func}(('foo', 'bar')) and x"),
            ("x.{func}('foo') or x.{func}('bar') or x", "x.{func}(('foo', 'bar')) or x"),
            
            # other name and/or on left
            ("y and x.{func}('foo') or x.{func}('bar')", "y and x.{func}(('foo', 'bar'))"),
            ("y or x.{func}('foo') or x.{func}('bar')", "y or x.{func}(('foo', 'bar'))"),
            
            # other name and/or on right
            ("x.{func}('foo') or x.{func}('bar') and y", "x.{func}(('foo', 'bar')) and y"),
            ("x.{func}('foo') or x.{func}('bar') or y", "x.{func}(('foo', 'bar')) or y"),

            # same name and/or on left, other name and/or on right
            ("x or x.{func}('foo') or x.{func}('bar') or y", "x or x.{func}(('foo', 'bar')) or y"),
            ("x and x.{func}('foo') or x.{func}('bar') or y", "x and x.{func}(('foo', 'bar')) or y"),

            # other name and/or on left, same name and/or on right
            ("y or x.{func}('foo') or x.{func}('bar') or x", "y or x.{func}(('foo', 'bar')) or x"),
            ("y and x.{func}('foo') or x.{func}('bar') or x", "y and x.{func}(('foo', 'bar')) or x"),
        ],
    )
    def test_mixed_boolean_operation(self, tmpdir, func, input_code, expected):
        self._format_func_run_test(tmpdir, func, input_code, expected)    
```

Reply:
Awesome! You can add it to this PR or create an issue with all this info to be worked on later

Reply:
I implemented a cleaner solution last night that doesn't require using the recursive func to extract operands and handles all cases regardless of surrounding boolean operands (all previous tests, test above, some new no_change tests). This is the version for startwiths, I'll update the PRs once I generalize and dedupe shared logic from both mods.

`combine_calls` handles the logic that was formerly in `make_new_call_from_boolean_operation` so I will generalize with the same idea you had in the previous message.

```python3
class CombineStartswithEndswith(SimpleCodemod, NameResolutionMixin):
    metadata = Metadata(
        name="combine-startswith-endswith",
        summary="Simplify Boolean Expressions Using `startswith` and `endswith`",
        review_guidance=ReviewGuidance.MERGE_WITHOUT_REVIEW,
        references=[],
    )
    change_description = "Use tuple of matches instead of boolean expression"

    args = [
        m.Arg(
            value=m.Tuple()
            | m.SimpleString()
            | m.ConcatenatedString()
            | m.FormattedString()
            | m.Name()
        )
    ]

    startswith_call_matcher = m.Call(
        func=m.Attribute(value=m.Name(), attr=m.Name("startswith")),
        args=args,
        )
    endswith_call_matcher = m.Call(
        func=m.Attribute(value=m.Name(), attr=m.Name("endswith")),
        args=args,
    )

    
    def leave_BooleanOperation(
        self, original_node: cst.BooleanOperation, updated_node: cst.BooleanOperation
    ) -> cst.CSTNode:
        if not self.filter_by_path_includes_or_excludes(
            self.node_position(original_node)
        ):
            return updated_node
        
        for call_matcher in (self.startswith_call_matcher, self.endswith_call_matcher):
            if self.matches_call_or_call(updated_node, call_matcher):
                self.report_change(original_node)
                return self.combine_calls(updated_node.left, updated_node.right)
                        
            if self.matches_call_or_boolop(updated_node, call_matcher):
                self.report_change(original_node)
                return self.call_or_boolop_fold_right(updated_node)
            
            if self.matches_boolop_or_call(updated_node, call_matcher):
                self.report_change(original_node)
                return self.call_or_boolop_fold_left(updated_node)

        return updated_node
    

    def matches_call_or_call(
        self, node: cst.BooleanOperation, call_matcher: m.Call
    ) -> bool:
        # Check for simple case: x.startswith("...") or x.startswith("...")
        call_or_call = m.BooleanOperation(left=call_matcher, operator=m.Or(), right=call_matcher)
        return (m.matches(node, call_or_call) # Same Func
            and node.left.func.value.value == node.right.func.value.value) # Same Instance


    def matches_call_or_boolop(
        self, node: cst.BooleanOperation, call_matcher: m.Call) -> bool:

        # Check for case when call on left and call on left of right boolop can be combined like:
        # x.startswith("...") or x.startswith("...") and/or <any>
        call_or_boolop = m.BooleanOperation(
            left=call_matcher,
            operator=m.Or(),
            right=m.BooleanOperation(left=call_matcher),
        )
        return (m.matches(node, call_or_boolop) # Same Func
                and node.left.func.value.value == node.right.left.func.value.value ) # Same Instance


    def matches_boolop_or_call(
        self, node: cst.BooleanOperation, call_matcher: m.Call) -> bool:

        # Check for case when call on right and call on right of left boolop can be combined like:
        # <any> and/or x.startswith("...") or x.startswith("...")
        boolop_or_call = m.BooleanOperation(
            left=m.BooleanOperation(right=call_matcher),
            operator=m.Or(),
            right=call_matcher,
        )
        return (m.matches(node, boolop_or_call) # Same Func
                and node.left.right.func.value.value == node.right.func.value.value ) # Same Instance  


    def combine_calls(self, *calls: cst.Call) -> cst.Call:
        elements = []
        seen_evaluated_values = set()
        for call in calls:
            arg_value = call.args[0].value
            arg_elements = (
                arg_value.elements
                if isinstance(arg_value, cst.Tuple)
                else (cst.Element(value=arg_value),)
            )

            for element in arg_elements:
                if (
                    evaluated_value := getattr(element.value, "evaluated_value", None)
                ) in seen_evaluated_values:
                    # If an element has a non-None evaluated value that has already been seen, continue to avoid duplicates
                    continue
                if evaluated_value is not None:
                    seen_evaluated_values.add(evaluated_value)
                elements.append(element)

        new_arg = cst.Arg(value=cst.Tuple(elements=elements))
        return cst.Call(func=call.func, args=[new_arg])
    

    def call_or_boolop_fold_right(
        self, node: cst.BooleanOperation
    ) -> cst.BooleanOperation:
        new_left = self.combine_calls(node.left, node.right.left)
        new_right = node.right.right
        return cst.BooleanOperation(left=new_left, operator=node.right.operator, right=new_right)


    def call_or_boolop_fold_left(
        self, node: cst.BooleanOperation
    ) -> cst.BooleanOperation:
        new_left = node.left.left
        new_right = self.combine_calls(node.left.right, node.right)
        return cst.BooleanOperation(left=new_left, operator=node.left.operator, right=new_right)
```